CFE-2977: Added sys.bootstrap_id policy variable

olehermanse opened the pull request:
Contains the ID from `/var/cfengine/bootstrap_id.dat`, if present.

olehermanse commented:
@cf-bottom: test in jenkins with exotics please.

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=2714)](https://ci.cfengine.com//job/pr-pipeline/2714/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2714/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2714/

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=2716)](https://ci.cfengine.com//job/pr-pipeline/2716/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2716/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2716/

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=2717)](https://ci.cfengine.com//job/pr-pipeline/2717/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2717/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2717/

olehermanse commented:
Tested, works:

```
root@dev core $ cf-agent --show-evaluated-vars | grep bootstrap
default:host_info_report_inventory.inventory_vars {"default:cfe_autorun_inventory_dmidecode.total_physical_memory_MB":"","default:cfe_autorun_inventory_ipv4_addresses.ipv4[10.0.2.15]":"10.0.2.15","default:cfe_autorun_inventory_ipv4_addresses.ipv4[192.168.100.10]":"192.168.100.10","default:cfe_autorun_inventory_listening_ports.ports":["22","111","5308","5432","34628","43716"],"default:inventory_any.id":"bootstrap","default:inventory_os.description":"ubuntu_14 (LSB missing)","default:sys.arch":"x86_64","default:sys.cf_version":"3.14.0a.7cd80e4","default:sys.class":"linux","default:sys.cpus":"2","default:sys.flavor":"ubuntu_14","default:sys.fqhost":"dev","default:sys.hardware_addresses":["08:00:27:74:61:f3","08:00:27:4d:5e:5a"],"default:sys.host":"dev","default:sys.interfaces":["eth0","eth1"],"default:sys.ipv4":"10.0.2.15","default:sys.key_digest":"MD5=381914d4b49b3d32b3077f01db8922f6","default:sys.os_release":{"BUG_REPORT_URL":"http://bugs.launchpad.net/ubuntu/","HOME_URL":"http://www.ubuntu.com/","ID":"ubuntu","ID_LIKE":"debian","NAME":"Ubuntu","PRETTY_NAME":"Ubuntu 14.04.5 LTS","SUPPORT_URL":"http://help.ubuntu.com/","VERSION":"14.04.5 LTS, Trusty Tahr","VERSION_ID":"14.04"},"default:sys.release":"3.13.0-161-generic","default:sys.uptime":"26","default:sys.uqhost":"dev"} source=promise
default:inventory_any.data               {"releaseId":"bootstrap"}                                    source=promise
default:inventory_any.id                 bootstrap                                                    source=promise,inventory,attribute_name=Policy Release Id
default:sys.bootstrap_id                 2p8senW5JylH0X7HsJnqEO1a3S7CY5o+rpv/kvu+                     source=bootstrap
default:sys.policy_hub                   192.168.100.10                                               source=bootstrap
default:sys.policy_hub_port              5308                                                         source=bootstrap
root@dev core $ cat /var/cfengine/bootstrap_id.dat
2p8senW5JylH0X7HsJnqEO1a3S7CY5o+rpv/kvu+
root@dev core $ echo blah >> /var/cfengine/bootstrap_id.dat
root@dev core $ cf-agent --show-evaluated-vars | grep bootstrap
default:host_info_report_inventory.inventory_vars {"default:cfe_autorun_inventory_dmidecode.total_physical_memory_MB":"","default:cfe_autorun_inventory_ipv4_addresses.ipv4[10.0.2.15]":"10.0.2.15","default:cfe_autorun_inventory_ipv4_addresses.ipv4[192.168.100.10]":"192.168.100.10","default:cfe_autorun_inventory_listening_ports.ports":["22","111","5308","5432","34628","43716"],"default:inventory_any.id":"bootstrap","default:inventory_os.description":"ubuntu_14 (LSB missing)","default:sys.arch":"x86_64","default:sys.cf_version":"3.14.0a.7cd80e4","default:sys.class":"linux","default:sys.cpus":"2","default:sys.flavor":"ubuntu_14","default:sys.fqhost":"dev","default:sys.hardware_addresses":["08:00:27:74:61:f3","08:00:27:4d:5e:5a"],"default:sys.host":"dev","default:sys.interfaces":["eth0","eth1"],"default:sys.ipv4":"10.0.2.15","default:sys.key_digest":"MD5=381914d4b49b3d32b3077f01db8922f6","default:sys.os_release":{"BUG_REPORT_URL":"http://bugs.launchpad.net/ubuntu/","HOME_URL":"http://www.ubuntu.com/","ID":"ubuntu","ID_LIKE":"debian","NAME":"Ubuntu","PRETTY_NAME":"Ubuntu 14.04.5 LTS","SUPPORT_URL":"http://help.ubuntu.com/","VERSION":"14.04.5 LTS, Trusty Tahr","VERSION_ID":"14.04"},"default:sys.release":"3.13.0-161-generic","default:sys.uptime":"27","default:sys.uqhost":"dev"} source=promise
default:inventory_any.data               {"releaseId":"bootstrap"}                                    source=promise
default:inventory_any.id                 bootstrap                                                    source=promise,inventory,attribute_name=Policy Release Id
default:sys.bootstrap_id                 2p8senW5JylH0X7HsJnqEO1a3S7CY5o+rpv/kvu+                     source=bootstrap
default:sys.policy_hub                   192.168.100.10                                               source=bootstrap
default:sys.policy_hub_port              5308                                                         source=bootstrap
root@dev core $ cat blah > /var/cfengine/bootstrap_id.dat
cat: blah: No such file or directory
root@dev core $ echo blah > /var/cfengine/bootstrap_id.dat
root@dev core $ cf-agent --show-evaluated-vars | grep bootstrap
cf-agent: bootstrap.c:409: ReadBootstrapIDFile: Assertion `data_length == (4 * (240 / 8 / 3)) + 1' failed.
root@dev core $ rm /var/cfengine/bootstrap_id.dat
root@dev core $ cf-agent --show-evaluated-vars | grep bootstrap
default:host_info_report_inventory.inventory_vars {"default:cfe_autorun_inventory_dmidecode.total_physical_memory_MB":"","default:cfe_autorun_inventory_ipv4_addresses.ipv4[10.0.2.15]":"10.0.2.15","default:cfe_autorun_inventory_ipv4_addresses.ipv4[192.168.100.10]":"192.168.100.10","default:cfe_autorun_inventory_listening_ports.ports":["22","111","5308","5432","34628","43716"],"default:inventory_any.id":"bootstrap","default:inventory_os.description":"ubuntu_14 (LSB missing)","default:sys.arch":"x86_64","default:sys.cf_version":"3.14.0a.7cd80e4","default:sys.class":"linux","default:sys.cpus":"2","default:sys.flavor":"ubuntu_14","default:sys.fqhost":"dev","default:sys.hardware_addresses":["08:00:27:74:61:f3","08:00:27:4d:5e:5a"],"default:sys.host":"dev","default:sys.interfaces":["eth0","eth1"],"default:sys.ipv4":"10.0.2.15","default:sys.key_digest":"MD5=381914d4b49b3d32b3077f01db8922f6","default:sys.os_release":{"BUG_REPORT_URL":"http://bugs.launchpad.net/ubuntu/","HOME_URL":"http://www.ubuntu.com/","ID":"ubuntu","ID_LIKE":"debian","NAME":"Ubuntu","PRETTY_NAME":"Ubuntu 14.04.5 LTS","SUPPORT_URL":"http://help.ubuntu.com/","VERSION":"14.04.5 LTS, Trusty Tahr","VERSION_ID":"14.04"},"default:sys.release":"3.13.0-161-generic","default:sys.uptime":"30","default:sys.uqhost":"dev"} source=promise
default:inventory_any.data               {"releaseId":"bootstrap"}                                    source=promise
default:inventory_any.id                 bootstrap                                                    source=promise,inventory,attribute_name=Policy Release Id
default:sys.policy_hub                   192.168.100.10                                               source=bootstrap
default:sys.policy_hub_port              5308                                                         source=bootstrap
root@dev core $
```

olehermanse commented:
Release build, bad cases are handled gracefully, with an error:

```
root@dev core $ cf-agent -B 192.168.100.10
R: Bootstrapping from host '192.168.100.10' via built-in policy '/var/cfengine/inputs/failsafe.cf'
R: This host assumes the role of policy server
R: Updated local policy from policy server
R: Triggered an initial run of the policy
  notice: Bootstrap to '192.168.100.10' completed successfully!
root@dev core $ cf-agent --show-evaluated-vars | grep bootstrap
default:sys.bootstrap_id                 wxHEJgX0DLPaijt1yquyVRcnmzrxZrlgN/XZn6ZK                     source=bootstrap
default:sys.policy_hub                   192.168.100.10                                               source=bootstrap
default:sys.policy_hub_port              5308                                                         source=bootstrap
root@dev core $ echo "blah" > /var/cfengine/bootstrap_id.dat
root@dev core $ cf-agent --show-evaluated-vars | grep bootstrap
   error: '/var/cfengine/bootstrap_id.dat' contains invalid data: 'blah
default:sys.policy_hub                   192.168.100.10                                               source=bootstrap
default:sys.policy_hub_port              5308                                                         source=bootstrap
root@dev core $ rm /var/cfengine/bootstrap_id.dat
root@dev core $ cf-agent --show-evaluated-vars | grep bootstrap
default:sys.policy_hub                   192.168.100.10                                               source=bootstrap
default:sys.policy_hub_port              5308                                                         source=bootstrap
root@dev core $ cf-agent -B 192.168.100.10
R: Bootstrapping from host '192.168.100.10' via built-in policy '/var/cfengine/inputs/failsafe.cf'
R: This host assumes the role of policy server
R: Updated local policy from policy server
R: Triggered an initial run of the policy
  notice: Bootstrap to '192.168.100.10' completed successfully!
root@dev core $ cf-agent --show-evaluated-vars | grep bootstrap
default:sys.bootstrap_id                 VbSEvpvMDpa57m/zZ9BbVZsC8E9ApgugNPAS4VIf                     source=bootstrap
default:sys.policy_hub                   192.168.100.10                                               source=bootstrap
default:sys.policy_hub_port              5308                                                         source=bootstrap
root@dev core $ cf-agent --show-evaluated-vars | grep bootstrap
default:sys.bootstrap_id                 VbSEvpvMDpa57m/zZ9BbVZsC8E9ApgugNPAS4VIf                     source=bootstrap
default:sys.policy_hub                   192.168.100.10                                               source=bootstrap
default:sys.policy_hub_port              5308                                                         source=bootstrap
root@dev core $ cf-agent -B 192.168.100.10
R: Bootstrapping from host '192.168.100.10' via built-in policy '/var/cfengine/inputs/failsafe.cf'
R: This host assumes the role of policy server
R: Updated local policy from policy server
R: Triggered an initial run of the policy
  notice: Bootstrap to '192.168.100.10' completed successfully!
root@dev core $ cf-agent --show-evaluated-vars | grep bootstrap
default:sys.bootstrap_id                 uVGv+Wj8YRa69RfYw7hR04IN6r0auJBdyzzMcAhY                     source=bootstrap
default:sys.policy_hub                   192.168.100.10                                               source=bootstrap
default:sys.policy_hub_port              5308                                                         source=bootstrap
root@dev core $ cf-agent --show-evaluated-vars | grep bootstrap
default:sys.bootstrap_id                 uVGv+Wj8YRa69RfYw7hR04IN6r0auJBdyzzMcAhY                     source=bootstrap
default:sys.policy_hub                   192.168.100.10                                               source=bootstrap
default:sys.policy_hub_port              5308                                                         source=bootstrap
root@dev core $
```

cf-bottom commented:
Sure, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=2718)](https://ci.cfengine.com//job/pr-pipeline/2718/)

(with exotics)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2718/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2718/

vpodzime commented on travis-scripts/valgrind.sh:
```diff
@@ -158,6 +158,14 @@ echo "Running cf-check:"
+diff id_a id_b
+
+echo "Checking that bootstrap ID has expected length"
+[ `cat id_a | awk '{print $2}' | wc -c` -eq 41 ]
```
Maybe it's time to rename the script and job? :wink:

olehermanse commented:
@vpodzime the problems with the test seem resolved by a longer sleep, please re-review.

olehermanse commented:
@cf-bottom jenkins

cf-bottom commented:
Alright, I triggered a build:

[![Build Status](https://ci.cfengine.com//buildStatus/icon?job=pr-pipeline&build=2726)](https://ci.cfengine.com//job/pr-pipeline/2726/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2726/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2726/
